pex/atomic_directory.py

```diff
@@ -85,80 +84,80 @@ class Value(Enum.Value):
+    # exclusive blocking file lock.
+
     atomic_dir = AtomicDirectory(target_dir=target_dir)
     if atomic_dir.is_finalized():
         # Our work is already done for us so exit early.
         yield atomic_dir
         return
 
-    lock_fd = None  # type: Optional[int]
+    head, tail = os.path.split(atomic_dir.target_dir)
+    if head:
+        safe_mkdir(head)
+
+    # N.B.: We don't actually write anything to the lock file but the fcntl file locking
+    # operations only work on files opened for at least write.
+    lock_fd = os.open(
+        os.path.join(head, ".{}.atomic_directory.lck".format(tail or "here")),
+        os.O_CREAT | os.O_WRONLY,
+    )
+
     lock_api = cast(
         "Callable[[int, int], None]",
-        fcntl.flock if exclusive is FileLockStyle.BSD else fcntl.lockf,
+        fcntl.flock if lock_style is FileLockStyle.BSD else fcntl.lockf,
     )
 
     def unlock():
         # type: () -> None
-        if lock_fd is None:
-            return
         try:
             lock_api(lock_fd, fcntl.LOCK_UN)
         finally:
             os.close(lock_fd)
 
-    if exclusive:
-        head, tail = os.path.split(atomic_dir.target_dir)
-        if head:
-            safe_mkdir(head)
-        # N.B.: We don't actually write anything to the lock file but the fcntl file locking
-        # operations only work on files opened for at least write.
-        lock_fd = os.open(
-            os.path.join(head, ".{}.atomic_directory.lck".format(tail or "here")),
-            os.O_CREAT | os.O_WRONLY,
-        )
-        # N.B.: Since lockf and flock operate on an open file descriptor and these are
-        # guaranteed to be closed by the operating system when the owning process exits,
-        # this lock is immune to staleness.
-        lock_api(lock_fd, fcntl.LOCK_EX)  # A blocking write lock.
-        if atomic_dir.is_finalized():
-            # We lost the double-checked locking race and our work was done for us by the race
-            # winner so exit early.
-            try:
-                yield atomic_dir
-            finally:
-                unlock()
-            return
+    # N.B.: Since lockf and flock operate on an open file descriptor and these are
+    # guaranteed to be closed by the operating system when the owning process exits,
+    # this lock is immune to staleness.
+    lock_api(lock_fd, fcntl.LOCK_EX)  # A blocking write lock.
+    if atomic_dir.is_finalized():
+        # We lost the double-checked locking race and our work was done for us by the race
+        # winner so exit early.
+        try:
+            yield atomic_dir
+        finally:
+            unlock()
+        return
 
+    # If there is an error making the work_dir that means file-locking guarantees have failed
+    # somehow and another process has the lock and has made the work_dir already. We let the error
+    # from os.mkdir propagate in that case.
+    os.mkdir(atomic_dir.work_dir)
```

Comment:
It could also mean that something segfaulted or was sigkilled, such that it exited uncleanly without tearing down the working directory?

Reply:
Yes, that's true. I'll step back and think about this a bit.

Reply:
Ok, I just warn now to ensure we get stderr data but clean up stale workdirs to prevent a stop-in-your-tracks bug that requires manual intervention.